src/include/pka_cpu.h

```diff
@@ -80,14 +86,191 @@ static inline uint64_t pka_early_cpu_speed(void)
+            goto err;
+        }
+        else
+            rd_cnt += rd_st;
+    }
+
+    if (rd_cnt != *len) {
+        printf("\n EOF reached before expected length.\n");
+        goto err;
+    }
+    else
+        goto out;
+
+err:
+    free(f);
+    f = NULL;
+out:
+    if (close(fd) == -1)
+        printf("\n Closing the file failed.\n");
+
+    return f;
+}
+
+static int pka_cpu_checksum(const uint8_t *buf, size_t len)
+{
+    uint8_t sum = 0;
+    size_t  d;
+
+    for (d = 0; d < len; d++)
+         sum += buf[d];
+
+    return sum == 0;
+}
+
+static uint16_t pka_cpu_read_dmi_table(uint32_t len, uint16_t num,
+                               const char *memdev)
+{
+    uint8_t *buf, *data, *next;
+    struct   dmi_header h;
+    size_t   size;
+    int      i = 0;
+
+    if ((buf = pka_cpu_file_read(&size, memdev)) == NULL)
+        return 0;
+
+    len  = size;
+    data = buf;
+
+    while ((i < num || !num)
+        && data + 4 <= buf + len) {
+
+        /// Copy the header information
+        h.type   = data[0];
+        h.length = data[1];
+        h.handle = WORD(data + 2);
+        h.data   = data;
+
+        if (h.length < 4) {
+            printf("\n Invalid entry length, DMI table is broken.\n");
+            break;
+        }
+        i++;
+
+        if (h.type == 127)
+            break;
+
+        next = data + h.length;
+        while ((unsigned long)(next - buf + 1) < len
+            && (next[0] != 0 || next[1] != 0))
+                next++;
+
+        next += 2;
+
+        if ((unsigned long)(next - buf) > len) {
+            printf("\n Structure is truncated at %dth structure.\n", i);
+            break;
+        }
+
+        /// Processor information
+        if (h.type == 4) {
+            if (h.length < 0x1A) {
+                printf("\n Incomplete processor information.\n");
+                return 0;
+            }
+            /// Frequency is at 0x16 offset
+            return WORD(h.data+0x16);
+        }
+
+        data = next;
+    }
+    return 0;
+}
+
+static uint64_t pka_cpu_get_frequency()
+{
+    uint16_t freq = 0;
+    uint8_t *buf;
+    size_t   size;
+
+    if ((buf = pka_cpu_file_read(&size,
+        "/sys/firmware/dmi/tables/smbios_entry_point")) != NULL) {
+        if (buf[0x06] > 0x20) {
+            printf("\n Entry point length exceeds the expected limit.\n");
+            return 0;
+        }
+
+        if (!pka_cpu_checksum(buf, buf[0x06])) {
+            printf("\n Checksum mismatch.\n");
+            return 0;
+        }
+
+        freq = pka_cpu_read_dmi_table(DWORD(buf + 0x0C), 0,
+            "/sys/firmware/dmi/tables/DMI");
+
+        return freq * MEGA;
+    }
+
+    return freq;
+}
+
 /// Returns maximum frequency of specified CPU (in Hz) on success, and 0
 /// on failure
 static inline uint64_t pka_cpu_hz_max_id(int id)
 {
-    if (id >= 0 && id < MAX_CPU_NUMBER)
-        return CPU_HZ_MAX;
+    if (id >= 0 && id < MAX_CPU_NUMBER) {

```

Comment:
Please reverse.

  if (id< 0 || id > MAX_CPU_NUMBER)
      return 0;

  <code to derive core frequency goes here>